keps/sig-node/3085-pod-conditions-for-starting-completition-of-sandbox-creation/README.md

```diff
@@ -1,1618 +1,1618 @@
+ of most pods (if the status manager does not batch them with other pod status
+ updates): one when pod sandbox creation completes and another when the pod is
+ terminated. However, there could be a higher number of patch calls to API Server
+ if the pod sandbox environment (like a microvm) starts successfully and then
+ crashes in a re-start loop.
+ 
+ Caching of updates to pod status by the pod status manager and batching pod
+ status updates (which is already in place) can help mitigate frequent patch
+ calls to API server.
+ 
+ ## Alternatives
+ 
+ <!--
+ What other approaches did you consider, and why did you rule them out? These do
+ not need to be as detailed as the proposal, but should include enough
+ information to express the idea and why it was not acceptable.
+ -->
+ 
+ ### Dedicated fields or annotations for the pod sandbox creation timestamps
+ 
+ Timestamps around completion of pod sandbox creation may be surfaced as a
+ dedicated field in the pod status rather than a pod condition. However, since
+ the successful creation of pod sandbox is essentially a "milestones" in the life
+ of a pod (similar to Scheduled, Ready, etc), pod conditions is the ideal place
+ to surface these and aligns well with the existing conditions like
+ `ContainersReady` and overall `Ready`.
+ 
+ A dedicated annotation on the pod for surfacing this data is another potential
+ approach. However, usage of annotations for Kubelet managed data is typically
+ discouraged.
+ 
+ ### Surface pod sandbox creation latency instead of timestamps
+ 
+ Surfacing the amount of time it took to successfully create a pod sandbox is an
+ alternative to surfacing the condition around completion of pod
+ sandbox creation (whose delta from pod scheduled condition reflects the
+ latency). The latency data would surface the same information from a pod
+ initialization SLI perspective as mentioned in the Motivations section.
+ Implementing this approach would require an API change on the pod status to
+ surface the latency data (as this no longer fits the structure of a pod
+ condition). This data cannot be consumed by other controllers as mentioned in
+ User Stories section.
+ 
+ ### Report sandbox creation latency as an aggregated metric
+ 
+ The duration it took pod sandbox to become ready can be directly reported as a
+ prometheus metrics aggregated in a histogram. However, aggregating the data at
+ the Kubelet level prevents a metric collection service from classifying the data
+ based on interesting fields on a pod (runtime class, storage class of PVCs,
+ number of PVCs, etc) or using custom labels and annotations on pods that
+ indicate workload characteristics (that the cluster operator may wish to use as
+ a basis for aggregating the metrics).
+ 
+ This also prevents other controllers from acting on sandbox status as
+ mentioned in User Stories section.
+ 
+ ### Report sandbox creation stages using Kubelet tracing
```

Comment:
Would something along these lines be viable?  If we continued to add monitoring of pod sandbox latencies for other things over time, how far can we go? If we want more granular monitoring, would we be okay with having 10 or even 50 conditions?  My hunch is that some type of tracing approach, or some other way to export metrics data scales much better.